### PR TITLE
added support for `restart-policy` keys in v3

### PR DESCRIPTION
**Pull request by surajnarwade**
Resolves `restart_policy` from issue #644

### CODE REVIEW

File: script/test/fixtures/v3/output-k8s-full-example.json [hunk 1]
```diff
@@ -160,7 +160,7 @@
         }
       },
       "spec": {
-        "replicas": 1,
+        "replicas": 6,
```
Comment (cdrage) [inline script/test/fixtures/v3/output-k8s-full-example.json]: neither should this..

File: script/test/fixtures/v3/output-env-subs.json [hunk 2]
```diff
@@ -56,8 +56,7 @@
                 "image": "foo/bar:latest",
                 "env": [
                   {
-                    "name": "FOO",
-                    "value": "bar"
+                    "name": "FOO"
```
Comment (cdrage) [inline script/test/fixtures/v3/output-env-subs.json]: this test should not be changed

File: script/test/fixtures/v3/output-k8s-full-example.json (hunk 1 above)
Comment (surajnarwade) [inline script/test/fixtures/v3/output-k8s-full-example.json]: but there are `replicas=6` in docker-compose file, right ?

File: pkg/loader/compose/v3.go [hunk 3]
```diff
@@ -183,21 +183,21 @@ func dockerComposeToKomposeMapping(composeObject *types.Config) (kobject.Kompose
...
 		serviceConfig.Args = composeServiceConfig.Command
-
 		// This is a bit messy since we use yaml.MemStringorInt
 		// TODO: Refactor yaml.MemStringorInt in kobject.go to int64
 		// Since Deploy.Resources.Limits does not initialize, we must check type Resources before continuing
 		if (composeServiceConfig.Deploy.Resources != types.Resources{}) {
 			serviceConfig.MemLimit = libcomposeyaml.MemStringorInt(composeServiceConfig.Deploy.Resources.Limits.MemoryBytes)
 		}
+		//Handling restart-policy
+		serviceConfig.Restart = composeServiceConfig.Deploy.RestartPolicy.Condition
```
Comment (cdrage) [inline pkg/loader/compose/v3.go]: Please add a small comment here in regards to deploy.

Example:

```
// Here we handle all Docker Compose Deploy keys
```

So we can better organize from the rest of all the keys :)

File: script/test/fixtures/v3/output-env-subs.json (hunk 2 above)
Comment (surajnarwade) [inline script/test/fixtures/v3/output-env-subs.json]: but results are like this, so I changed

File: script/test/fixtures/v3/output-k8s-full-example.json [hunk 4]
```diff
@@ -363,13 +358,10 @@
                 "stdin": true,
                 "tty": true
               }
-            ],
-            "restartPolicy": "Always"
```
Comment (cdrage) [inline script/test/fixtures/v3/output-k8s-full-example.json]: and here

File: script/test/fixtures/v3/output-k8s-full-example.json [hunk 5]
```diff
@@ -152,11 +152,6 @@
         "creationTimestamp": null,
         "labels": {
           "io.kompose.service": "foo"
-        },
-        "annotations": {
-          "com.example.description": "Accounting webapp",
-          "com.example.empty-label": "",
-          "com.example.number": "42"
```
Comment (cdrage) [inline script/test/fixtures/v3/output-k8s-full-example.json]: Why are these being deleted?

File: script/test/fixtures/v3/output-k8s-full-example.json [hunk 6]
```diff
@@ -363,13 +358,10 @@
...
                 "tty": true
               }
-            ],
-            "restartPolicy": "Always"
+            ]
           }
         },
-        "strategy": {
-          "type": "Recreate"
-        }
```
Comment (cdrage) [inline script/test/fixtures/v3/output-k8s-full-example.json]: Same as here?

File: script/test/fixtures/v3/output-os-full-example.json [hunk 7]
```diff
@@ -152,16 +152,10 @@
         "creationTimestamp": null,
         "labels": {
           "io.kompose.service": "foo"
-        },
-        "annotations": {
-          "com.example.description": "Accounting webapp",
-          "com.example.empty-label": "",
-          "com.example.number": "42"
```
Comment (cdrage) [inline script/test/fixtures/v3/output-os-full-example.json]: Annotations should still work ^^

File: script/test/fixtures/v3/output-k8s-full-example.json (hunk 5 above)
Comment (surajnarwade) [inline script/test/fixtures/v3/output-k8s-full-example.json]: @cdrage, they are not deleted, just order is misplaces

File: script/test/fixtures/v3/output-os-full-example.json (hunk 7 above)
Comment (surajnarwade) [inline script/test/fixtures/v3/output-os-full-example.json]: same here

**Comment by surajnarwade** [comment]
@cdrage , there was typo ( `on_failure` >> `on-failure`) in example due to which output was varying, can you please review now.

File: pkg/loader/compose/v3.go [hunk 8]
```diff
@@ -183,22 +183,24 @@ func dockerComposeToKomposeMapping(composeObject *types.Config) (kobject.Kompose
...
-
 		// This is a bit messy since we use yaml.MemStringorInt
 		// TODO: Refactor yaml.MemStringorInt in kobject.go to int64
 		// Since Deploy.Resources.Limits does not initialize, we must check type Resources before continuing
 		if (composeServiceConfig.Deploy.Resources != types.Resources{}) {
 			serviceConfig.MemLimit = libcomposeyaml.MemStringorInt(composeServiceConfig.Deploy.Resources.Limits.MemoryBytes)
 		}
-
+		//Here we handle all Docker Compose Deploy keys
+		//Handling restart-policy
```
Comment (cdrage) [inline pkg/loader/compose/v3.go]: please add spaces after `//`

File: pkg/loader/compose/v3.go [hunk 9]
```diff
@@ -183,22 +183,24 @@ func dockerComposeToKomposeMapping(composeObject *types.Config) (kobject.Kompose
...
+		if composeServiceConfig.Deploy.RestartPolicy != nil {
```
Comment (cdrage) [inline pkg/loader/compose/v3.go]: no need for `!= nil`, just put: `if composeServiceConfig.Deploy.RestartPolicy  {`

File: script/test/fixtures/v3/docker-compose-full-example.yaml [hunk 10]
```diff
@@ -39,7 +39,7 @@ services:
           cpus: '0.0001'
           memory: 20M
       restart_policy:
-        condition: on_failure
+        condition: on-failure
```
Comment (cdrage) [inline script/test/fixtures/v3/docker-compose-full-example.yaml]: unfortunatley, docker says that it works with either _ or -. :-1: but keep this change in!

File: pkg/transformer/kubernetes/k8sutils.go [hunk 11]
```diff
@@ -441,9 +441,9 @@ func (k *Kubernetes) UpdateKubernetesObjects(name string, service kobject.Servic
 
 		// Configure the container restart policy.
 		switch service.Restart {
-		case "", "always":
+		case "", "always", "any":
 			template.Spec.RestartPolicy = api.RestartPolicyAlways
-		case "no":
+		case "no", "none":
 			template.Spec.RestartPolicy = api.RestartPolicyNever
 		case "on-failure":
```
Comment (cdrage) [inline pkg/transformer/kubernetes/k8sutils.go]: will actually have to add `on_failure` as well to here!

Comment (surajnarwade) [inline pkg/transformer/kubernetes/k8sutils.go]: but https://docs.docker.com/compose/compose-file/#restart_policy shows `on-failure`

Comment (surajnarwade) [inline pkg/transformer/kubernetes/k8sutils.go]: `on-failure` case is already there ?

Comment (cdrage) [inline pkg/transformer/kubernetes/k8sutils.go]: I know, but on_failure is still supported. I can't remember where in `docker/cli` is it, but it's there. It automatically converts. Unfortunately, some examples use `on_failure`.

**Comment by surajnarwade** [comment]
@cdrage , `on_failure` case added, final review please and ready for merge :)

File: pkg/transformer/kubernetes/k8sutils.go [hunk 12]
```diff
@@ -441,11 +441,11 @@ func (k *Kubernetes) UpdateKubernetesObjects(name string, service kobject.Servic
...
 		// Configure the container restart policy.
 		switch service.Restart {
-		case "", "always":
+		case "", "always", "any":
 			template.Spec.RestartPolicy = api.RestartPolicyAlways
-		case "no":
+		case "no", "none":
 			template.Spec.RestartPolicy = api.RestartPolicyNever
-		case "on-failure":
+		case "on-failure", "on_failure":
```
Comment (cdrage) [inline pkg/transformer/kubernetes/k8sutils.go]: so found out it was error upstream. it's fixed now :( so we can remove `on_failure`. Sorry for confusion!

https://github.com/docker/cli/pull/262

File: script/test/fixtures/v3/docker-compose-full-example.yaml (hunk 10 above)
Comment (cdrage) [inline script/test/fixtures/v3/docker-compose-full-example.yaml]: fixed: https://github.com/docker/cli/pull/262

**Comment by surajnarwade** [comment]
@kadel @surajssd  needed one more +1 here :)

File: pkg/transformer/kubernetes/k8sutils.go (hunk 12 above)
Comment (cdrage) [inline pkg/transformer/kubernetes/k8sutils.go]: @surajnarwade gotta remove on_failure and we're good :+1:

**Comment by surajnarwade** [comment]
@cdrage , removed `on_failure`, we are good to go

**Comment by cdrage** [comment]
@surajnarwade thanks for helping me in terms of the confusion! this LGTM

**Comment by surajnarwade** [comment]
@surajssd , needs +1 to get it merged

**Pull request merged**